Upgrade on communication protocol

gimait opened the pull request:
This PR should add the functionality to form frames including several packets and resending packets that have not been acknowledged by the robot.

The first commit has not been tested yet, please check that the idea behind the changes makes sense!

zayfod commented on pycozmo/conn.py:
```diff
@@ -76,26 +80,32 @@ def run(self) -> None:
-            self.handle_pkt(pkt)
+            frame_trusted = self.handle_pkt(pkt)
+            if not frame_trusted:
+                return
```
No need to return here. The stream of packets is independent from the acknowledgement, which is for the stream of packets in the other direction

```
          ------- packets ------>
          <-- acknowledgments ---
engine                              robot
          <------ packets -------
          --- acknowledgments -->
```

gimait replied:
You are right, sorry, I think I meant to have that with one indent less, so skip the acknowledgement if the package is repeated.

I think I will also include the logic to manage the acknowledgement in the send window. I only avoided it because it means to include a new variable to track the progress of the acknowledgement (and I felt lazy about it I guess haha). This is necessary because the robot sends an accumulate acknowledgement (I think it's called like that), which does not match the address/id of the packets received, but rather counts how many have been received in order.

zayfod commented on pycozmo/conn.py:
```diff
@@ -158,19 +168,35 @@ def run(self) -> None:
+                pkts = []
+                for p in pkts:
+                    framelen += len(p) + 1
+                    if framelen < 1000:
```
We should define `MAX_FRAME_SIZE` which will probably be something like 1492 - `MIN_FRAME_SIZE` = 1478.

I'll try to check in the PCAPs that I have what is the maximum frame size, I've seen between Cozmo and the app.

gimait replied:
Completely agree, I put a random 1000 based on the biggest frames I've seen so far, so I'd be happy if you can find out the max size we can use.

zayfod replied:
I checked, and the maximum frame size I've seen in PCAPs is 1420 B.

This is the total size of the UDP packet data, including the Cozmo frame header.

zayfod commented on pycozmo/conn.py:
```diff
@@ -158,19 +168,35 @@ def run(self) -> None:
-                else:
-                    frame = Frame(protocol_declaration.FrameType.ENGINE, seq, seq, self.last_ack, [pkt])
-            raw_frame = frame.to_bytes()
+                first_seq, pkts = self.window.put(pkt)
```
Instead of modifying the return value of `put()`, why not accessing `expected_seq` and `window` directly from `SendThread`?

gimait replied:
I guess just to keep a bit of the previous implementation and keep it into one line, but I agree that it would be clearer to separate the put method from the access to the current window and sequence.

zayfod commented on pycozmo/conn.py:
```diff
@@ -158,19 +168,35 @@ def run(self) -> None:
 
             try:
-                self.sock.sendto(raw_frame, self.receiver_address)
+                for raw_frame in raw_frames:
```
I love the fact that you are working on sending multiple packets with one `Frame`.

The current implementation though, I think will resend all unacknowledged packets on every new packets.

The goal instead should be to look into the `SendThread` queue, pack whatever is in it into frames and send these.

gimait replied:
Yep, I agree, that would be better. I can also include a special case for the ping so it's sent instantly instead of wait for the time to send the packets. We could have a while loop that empties the queue and sends the available packets on a timeout, but if a ping is received then we send it without waiting or affecting the timeout.

zayfod reviewed (commented):
The direction in which you are working is great!

Asking for a few changes:
- pack the content of `SendThread.queue` into multiple frames and not `SendThread.window`.
- a new resend mechanism is needed that will trigger the resending of unacknowledged packets from `SendThread.window` if after X milliseconds, an acknowledgement has not been received.
- let's try to minimize the communication between `SendThread` and `ReceiveThread` to one callback function e.g. `SendThread.acknowledge(seq)`. It can be called from `ReceiveThread` whenever a new frame is received and it will be the only one needing synchronization with a lock internally

Let me know what you think.

gimait commented on pycozmo/window.py:
```diff
@@ -98,44 +99,67 @@ def __init__(self, seq_bits: int, size: Optional[int] = None) -> None:
+                    for frame in self.window[self.expected_seq:]:
+                        frame.reset()
+                    for frame in self.window[:seq]:
+                        frame.reset()
```
You are right, something like this?
```
for i in range(seq + self.expected_seq - self.size):
    self.window[(i+self.expected_seq)%self.size].reset()
```
This could substitute lines 145 to 152 actually, much better, thanks!

gimait commented on pycozmo/window.py:
```diff
@@ -5,6 +5,7 @@
 """
 
 import math
+from threading import Lock
```
I agree, I also stopped liking to share the windows between send and receive threads after I started thinking on the thread safety.
I am thinking that, since the send thread does not need to look into the receive window, we could pass the send thread object as a member of receive thread, and keep the windows as they were before, and add a method in the send thread to do the needed interactions with the send window safely (as I think you mentioned somewhere).

gimait commented:
I updated the PR with the changes you suggested, and now it seems to work well. I tried it with some commands and audio and it looks like it works, but I wouldn't mind some rechecking and extra testing.

There is one thing that I'm having trouble thinking of a solution for, and maybe you know how to do it. You mentioned that
> a new resend mechanism is needed that will trigger the resending of unacknowledged packets from SendThread.window if after X milliseconds, an acknowledgement has not been received.

I tried to think on ways to do this, but I can't really think on a way to keep track of the acknowledgement times other than how it's done that doesn't mean adding threads or unnecessary complexity.

The way that is done now is rather simple:
- The send thread has a timer that generates and sends frames at a regular interval. 
- Inside the run's loop, there is another loop, checking for new packets. I updated the timeout for reading the queue so that it wouldn't affect greatly the timing of the new messages.
- The send window is filled with each new command (packet) to send, and when the time comes, all packets in the window are sent.
- The receive thread gets some messages back and sends the acknowledgement to the send window, that clears old messages 
 and updates the corresponding indexes.

So far in my tests, it never happened that I had more than one or two repeated packages at the time, but it would be good to test this with more complete applications, including animations and sound and see what happens.

Let me know if you see anything that needs change ;)

zayfod commented on pycozmo/protocol_declaration.py:
```diff
@@ -20,6 +21,7 @@
 
 FRAME_ID = b"COZ\x03RE\x01"
 MIN_FRAME_SIZE = len(FRAME_ID) + 1 + 2 + 2 + 2
+MAX_FRAME_SIZE = 1420 - MIN_FRAME_SIZE
```
A better name for this may be `MAX_FRAME_PAYLOAD_SIZE`. Currently `MIN_FRAME_SIZE` is the total size of a frame with no payload but `MAX_FRAME_SIZE` is the maximum payload size only (minus the header`.

Also, I noticed that the robot drops frames, larger than 14+1037=1051 bytes with the following message:
```
WiFi.Messages.ProcessMessage.TooBig: Received message too big! 97(1375) > 1037
```

It seems that this is the maximum payload size.

zayfod commented on pycozmo/window.py:
```diff
@@ -130,6 +130,28 @@ def pop(self) -> None:
+
+    def get(self):
+        expected_idx = self.expected_seq % self.size
+        next_idx = self.next_seq % self.size
```
No, need for `% self.size` here as this was done on assignment.

zayfod commented on pycozmo/conn.py:
```diff
@@ -106,6 +243,9 @@ def handle_pkt(self, pkt: Packet) -> None:
             self.deliver_sequence()
+            self.send_thread.set_last_recv_ack(pkt.seq)
+
+        return
```
No need for this return

zayfod commented on pycozmo/conn.py:
```diff
@@ -37,10 +38,144 @@
+
+            # if self.window.is_full():
+            #     time.sleep(self.timeout)
+            #     continue
```
Let's get rid of this. These are old resend experiments.

zayfod commented on pycozmo/conn.py:
```diff
@@ -37,10 +38,144 @@
+        self.queue = Queue()
+        self.queue_timeout = queue_timeout
+        self.last_ack = 1
+        self.last_send_timestamp = datetime.now() - timedelta(days=1)
```
👍

zayfod commented on pycozmo/conn.py:
```diff
@@ -37,10 +38,144 @@
+                for raw_frame in raw_frames:
+                    self._send_frame(raw_frame)
+
+                self.last_send_timestamp = datetime.now()
```
Will it make sense to you to do this from within `_send_frame()`?

zayfod commented on pycozmo/conn.py:
```diff
@@ -37,10 +38,144 @@
+                                          ).to_bytes()
+                        self._send_frame(raw_frame)
+                    else:
+                        with self.lock:
```
Lock is not needed as long as the window is used by the SendThread only.

zayfod commented on pycozmo/conn.py:
```diff
@@ -37,10 +38,144 @@
+                raw_frames = []
+                framelen = 0
+                first_seq, pkts, last_ack = None, None, None
+                with self.lock:
```
Lock is not needed as long as the window is used by the SendThread only.

zayfod commented on pycozmo/conn.py:
```diff
@@ -37,10 +38,144 @@
+            #     time.sleep(self.timeout)
+            #     continue
+
+            while datetime.now() - self.last_send_timestamp < timedelta(seconds=self.loop_timeout):
```
What is the purpose of this loop?

gimait commented:
I found a bug in the `window`, the max_size had to be reduced by 1, since to encode the integer value in the given bits you need to account for the 0 value (ends up being 2^n_bits - 1).

I think the code is good to go, but let me know if you would like me to apply some more changes to the window or the connection.
